libs/ui/src/unlock_machine_screen.tsx

```diff
@@ -39,43 +43,67 @@ const EnteredCode = styled.div`
 
 type CheckingPassCodeAuth =
   | DippedSmartcardAuth.CheckingPasscode
+  | DippedSmartCardAuth.CheckingPin
   | InsertedSmartcardAuth.CheckingPasscode;
 
 interface Props {
   auth: CheckingPassCodeAuth;
   grayBackground?: boolean;
+  checkPin?: (pin: string) => Promise<void>;
 }
 
 export function UnlockMachineScreen({
   auth,
   grayBackground,
+  checkPin: checkPinFromProps,
 }: Props): JSX.Element {
   assert(auth.status === 'checking_passcode');
 
   const [currentPasscode, setCurrentPasscode] = useState('');
-  const handleNumberEntry = useCallback((digit: number) => {
-    setCurrentPasscode((prev) =>
-      `${prev}${digit}`.slice(0, SECURITY_PIN_LENGTH)
-    );
-  }, []);
+
+  const handleNumberEntry = useCallback(
+    async (digit: number) => {
+      const passcode = `${currentPasscode}${digit}`.slice(
+        0,
+        SECURITY_PIN_LENGTH
+      );
+      setCurrentPasscode(passcode);
+      if (passcode.length === SECURITY_PIN_LENGTH) {
+        if ('checkPasscode' in auth) {
+          auth.checkPasscode(passcode);
+        } else {
+          assert(checkPinFromProps !== undefined);
+          await checkPinFromProps(passcode);
+        }
+        setCurrentPasscode('');
+      }
```

Comment:
Your favorite kind of change @jonahkagan! Moved PIN checking logic out of a `useEffect` and into the relevant event handler

Actually had to do this to address a bug where the PIN was being checked twice, the second check resulting in an error on the backend because we were no longer in the `CheckingPin` state